core/src/main/java/org/opensearch/sql/data/model/ExprDateValue.java

```diff
@@ -36,9 +43,44 @@ public ExprDateValue(String date) {
     }
   }
 
+  /** Constructor of ExprDateValue to support custom/OpenSearch date formats in mappings. */
+  public ExprDateValue(String date, List<DateFormatter> dateFormatters) {
```

Comment:
try to understand what we are tyring to achieve. Construct a ExprDateValue from string value, when call getValue, the result should be formatted with provided formatter? If yes, we should move the format logic to getValue instead of in constructor where the value is lazy evaluated.

Reply:
> I'm parsing date string using OpenSearch DateFormatters

Core module should not directly depend on OpenSearch library. This is why we have OpenSearch module.

Reply:
> There is no lazy variable evaluation in Java

we can simulate lazy evalution, right?

> especially there is a for loop in this logic.

another, could we avoid for...loop here?

Reply:
Instead change value() of ExprDateValue, format the ExprDateValue.value() based on Type. for instance
```
if (this.ref.type instance of OpenSearchDataType) {
  ((OpenSearchDataType) this.ref.type instance).getAllFormatters.format(literal.value())
}
```

Reply:
I see what you mean. Instead of bringing OpenSearch library dependency to `core`, I can do all the formatting logic within `OpenSearchDateType ` class which already has access to OpenSearch DateFormatters. 
```
if (this.ref.type instance of OpenSearchDataType) {
  ((OpenSearchDataType) this.ref.type instance).getAllFormatters.format(literal.value())
}
```
But the above logic cannot go inside `ExprDatevalue` class because I cannot import `OpenSearchDateType` class within core as it will make it circular dependency. (OpenSearch module already has dependency on core modules)

Synced offline with @penghuo to make the below changes:

- Move DateFormatter logic inside `OpenSearchDateType` class 
- Lucene query class refers to `OpenSearchDateType.getParsedDateTime()` function to get the parsed date object and pass it to `ExprDateValue` constructor
- Remove new constructors from ExprDateType and similar classes
- Formatted value will be fetched inside term query and range query referring to `OpenSearchDateType` class directly.